heracal/omni.py

```diff
@@ -877,15 +978,404 @@ def __init__(self, meta, gains, flags=None, DELAY=False, ex_ants=[], appendhist=
+        pp = getPol(f)
+        djd = file2djd(f)
+        if not opts.firstcal:
+            if len(pols) == 1:
+                # atomic solution application
+                # XXX this is fragile
+                fexpected = solution_files[i]
+            else:
+                # one solution file per djd
+                fexpected = next((s for s in solution_files if djd in s), None)
+            try:
+                ind = solution_files.index(fexpected)
+                filedict[f] = str(solution_files[ind])
+            except ValueError:
+                raise Exception(
+                    'Solution file %s expected; not found.' % fexpected)
+
+        else:
+            if nf == len(solution_files) * len(pols):  # atomic firstcal application
+                filedict[f] = solution_files[i]  # XXX this is fragile
+            else:  # one firstcal file for many data files
+                if isLinPol(pp):
+                    filedict[f] = [firstcal_files[pp][0]]
+                else:
+                    filedict[f] = [firstcal_files[lpk][0]
+                                   for lpk in linear_pol_keys]
+
+    for f in files:
+        mir = UVData()
+        print("  Reading {0}".format(f))
+        mir.read_miriad(f)
+        if mir.phase_type != 'drift':
+            mir.unphase_to_drift()
+        cal = UVCal()
+        print("  Reading calibration : {0}".format(filedict[f]))
+        if len(pols) == 1 or not opts.firstcal:
+            cal.read_calfits(filedict[f])
+        else:
+            if isLinPol(getPol(f)):
+                cal.read_calfits(filedict[f][0])
+            else:
+                cal = heracal.omni.concatenate_UVCal_on_pol(filedict[f])
+
+        print("  Calibrating...")
+        antenna_index = dict(zip(*(cal.ant_array, range(cal.Nants_data))))
+        for p, pol in enumerate(mir.polarization_array):
+            # XXX could replace with numpy function instead of casting to list
+            p1, p2 = [list(cal.jones_array).index(pk)
+                      for pk in jonesLookup[pol]]
+            for bl, k in zip(*np.unique(mir.baseline_array, return_index=True)):
+                blmask = np.where(mir.baseline_array == bl)[0]
+                ai, aj = mir.baseline_to_antnums(bl)
+                if not ai in cal.ant_array or not aj in cal.ant_array:
+                    continue
+                for nsp, nspws in enumerate(mir.spw_array):
+                    if cal.cal_type == 'gain' and cal.gain_convention == 'multiply':
+                        mir.data_array[blmask, nsp, :, p] = \
+                            mir.data_array[blmask, nsp, :, p] * \
+                            cal.gain_array[antenna_index[ai], nsp, :, :, p1].T * \
+                            np.conj(cal.gain_array[
+                                    antenna_index[aj], nsp, :, :, p2].T)
+
+                    if cal.cal_type == 'gain' and cal.gain_convention == 'divide':
+                        mir.data_array[blmask, nsp, :, p] =  \
```

Comment:
I don't like that we're forced to have four of these gain application brackets, but I don't think it's avoidable right now